homeassistant/components/tuya.py

```diff
@@ -0,0 +1,166 @@
+    dispatcher_send, async_dispatcher_connect)
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.event import track_time_interval
+
+REQUIREMENTS = ['tuyapy==0.1.1']
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_COUNTRYCODE = 'country_code'
+
+DOMAIN = 'tuya'
+DATA_TUYA = 'data_tuya'
+
+SIGNAL_DELETE_ENTITY = 'tuya_delete'
+SIGNAL_UPDATE_ENTITY = 'tuya_update'
+
+SERVICE_FORCE_UPDATE = 'force_update'
+SERVICE_PULL_DEVICES = 'pull_devices'
+
+TUYA_TYPE_TO_HA = {
+    'switch': 'switch'
+}
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_PASSWORD): cv.string,
+        vol.Required(CONF_USERNAME): cv.string,
+        vol.Required(CONF_COUNTRYCODE): cv.string
+    })
+}, extra=vol.ALLOW_EXTRA)
+
+
+def setup(hass, config):
+    """Set up Tuya Component."""
+    from tuyapy import TuyaApi
+
+    tuya = TuyaApi()
+    username = config[DOMAIN][CONF_USERNAME]
+    password = config[DOMAIN][CONF_PASSWORD]
+    country_code = config[DOMAIN][CONF_COUNTRYCODE]
+
+    hass.data[DATA_TUYA] = tuya
+    tuya.init(username, password, country_code)
+    hass.data[DOMAIN] = {
+        'entities': {}
+    }
+
+    def load_devices(device_list):
+        """Load new devices by device_list."""
+        device_type_list = {}
+        for device in device_list:
+            dev_type = device.device_type()
+            if (dev_type in TUYA_TYPE_TO_HA and
+                    device.object_id() not in hass.data[DOMAIN]['entities']):
+                ha_type = TUYA_TYPE_TO_HA[dev_type]
+                if ha_type not in device_type_list:
+                    device_type_list[ha_type] = []
+                device_type_list[ha_type].append(device.object_id())
+        for ha_type, dev_ids in device_type_list.items():
+            discovery.load_platform(
+                hass, ha_type, DOMAIN,
+                {'dev_ids': dev_ids}, config)
+
+    device_list = tuya.get_all_devices()
+    load_devices(device_list)
+
+    def poll_devices_update(event_time):
+        """Check if accesstoken is expired and pull device list from server."""
+        _LOGGER.info("Pull Devices From Tuya")
+        tuya.poll_devices_update()
+        # Add new discover device.
+        device_list = tuya.get_all_devices()
+        load_devices(device_list)
+        # Delete not exist device.
+        newlist_ids = []
+        for device in device_list:
+            newlist_ids.append(device.object_id())
+        for dev_id in list(hass.data[DOMAIN]['entities']):
+            if dev_id not in newlist_ids:
+                dispatcher_send(hass, SIGNAL_DELETE_ENTITY, dev_id)
+                hass.data[DOMAIN]['entities'].pop(dev_id)
+
+    track_time_interval(hass, poll_devices_update, timedelta(minutes=5))
+
+    hass.services.register(DOMAIN, SERVICE_PULL_DEVICES, poll_devices_update)
+
+    def force_update(call):
+        """Force all devices to pull data."""
+        _LOGGER.info("Refreshing Device Data From Tuya")
```

Comment:
The service call will already be logged at info level by the core, so I don't think this logging is needed.